### PR TITLE
Adapt `BigInt::random_integer()`

### LINKED ISSUE #109: BigInt::random_integer() needs adaption
This algorithm is specifically explained and referenced to the TR. It was [adapted in Botan 3.1.0](https://github.com/randombit/botan/pull/3594) though. Adapt the crypto doc.

### PR DESCRIPTION
**Pull request by reneme**
Closes #109

### CODE REVIEW

File: docs/cryptodoc/src/07_rng.rst [hunk 1]
```diff
@@ -1054,5 +1060,29 @@ It works as follows:
...
+**Note:** Technically, the base implementation above should be defined for exactly :math:`min = 0`
+not :math:`min \in \{0,1\}`. When `adding this special case in Botan 3.1.0 <https://github.com/randombit/botan/pull/3594>`_
+this was done to avoid breaking existing test cases that depended on the determinism of this method for :math:`min = 1`.
+
+.. admonition:: ``BigInt::random_integer()`` (for :math:`min > 1`)
+
+   **Input:**
+
+   -  ``rng``: random number generator (see :ref:`rng/main` for available algorithms)
+   -  ``min``: integer lower bound of desired range (with :math:`min > 1`)
```
Comment (FAlbertDev) [inline docs/cryptodoc/src/07_rng.rst:1072]: I think we should stay consistent with writing _min_, _max_, and _r_ in code style or in math style. (Not only here but in this whole section)

File: docs/cryptodoc/src/07_rng.rst [hunk 2]
```diff
@@ -1054,5 +1060,29 @@ It works as follows:
...
+   -  ``max``: integer upper bound of desired range, excluding ``max``
+
+   **Output:**
+
+   -  ``r``: :math:`min \leq r < max`
+
+   **Steps:**
+
+   1. Preliminary parameter requirement checks are conducted. ``min`` must be
+      smaller than ``max``, and neither can be negative.
```
Comment (FAlbertDev) [inline docs/cryptodoc/src/07_rng.rst:1082]: Not here, but above, for the algorithm with _min_ in {0,1}: I would expect a check like: _min_ in {0,1} and _min_ < _max_.

File: docs/cryptodoc/src/07_rng.rst [hunk 3]
```diff
@@ -1054,5 +1060,29 @@ It works as follows:
...
+   2. Calculate ``diff`` as the difference between ``max`` and ``min``
+   3. Call ``BigInt::random_integer()`` for the interval ``0..diff`` to obtain ``n``
```
Comment (FAlbertDev) [inline docs/cryptodoc/src/07_rng.rst:1084]: in the interval `0..diff`: `diff` is excluded. Maybe we want to write something like:

Call ``BigInt::random_integer()`` to obtain a random number _n_ with _0 <= n < diff_

File: docs/cryptodoc/src/07_rng.rst [hunk 4]
```diff
@@ -1054,5 +1060,29 @@ It works as follows:
    4. if (:math:`min \leq r < max`) return ``r``.
```
Comment (FAlbertDev) [inline docs/cryptodoc/src/07_rng.rst:1060]:
```suggestion
   4. If (:math:`min \leq r < max`) return ``r``.
```

**Review by FAlbertDev — COMMENTED** [review]
Some small comments, but LGTM in general. 
It feels a bit wrong to "hack" the case for min = 1 within the algorithm for min = 0. Maybe we should investigate what expense is necessary to update the respective tests so that they also work with the second algorithm.

File: docs/cryptodoc/src/07_rng.rst [hunk 5]
```diff
@@ -1031,14 +1031,20 @@ generation of random parameters for DSA signatures.
...
 In order to generate integers from an arbitrary range as required for the implemented public key algorithms,
 Botan uses the ``BigInt::random_integer()`` method (implemented in :srcref:`src/lib/math/bigint/big_rand.cpp`).
-It works as follows:
+Note that this algorithm allows to generate random numbers from an arbitrary range
+between a ``min`` and a ``max`` (exclusive) value. Typically, in public key algorithms,
+the ``min`` parameter will be zero.
 
-.. admonition:: ``BigInt::random_integer()``
+The algorithm has a special case for :math:`min > 1` to avoid unpredictable runtimes
+if ``min`` and ``max`` are both large but their difference is small.
```
[the remark this answers is not in the document]
Comment (reneme) [reply docs/cryptodoc/src/07_rng.rst:1039]: [Grammarly says](https://www.grammarly.com/blog/comma-before-but/):

> No comma with but: when but is used to connect phrases but not independent clauses

The second phrase refers to the first and can't really stand on its own. I believe there shouldn't be a comma.

Meta: Thanks for pointing these things out. I place commas purely by gut-feeling and it's great to be inspired to look at the rules every now and then. 😏

File: docs/cryptodoc/src/07_rng.rst (hunk 3 above)
Comment (reneme) [reply docs/cryptodoc/src/07_rng.rst:1084]: Making it:

```
3. Call ``BigInt::random_integer(rng, 0, diff)`` to obtain :math:`n`
```

No need to overly abstract that. I code we trust.

File: docs/cryptodoc/src/07_rng.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (FAlbertDev) [reply docs/cryptodoc/src/07_rng.rst]: Well, I checked the text with Grammarly, and it marked the missing comma. Also, I would say the phrase "Their difference is small." could stand on its own and matches more with the other examples with commas at [Grammarly's explanation page](https://www.grammarly.com/blog/comma-before-but/). 
Grammar Police: Over and out.

File: docs/cryptodoc/src/07_rng.rst (hunk 2 above)
Comment (FAlbertDev) [reply docs/cryptodoc/src/07_rng.rst]: Alright!

**Comment by FAlbertDev** [comment]
> Interestingly, even the RFC8448 tests depend on this RNG strategy.

Interesting... We would need to manually find a new fixed rng input vector that generates the right values in the right positions, correct? This doesn't sound like fun...

File: docs/cryptodoc/src/07_rng.rst [hunk 6]
```diff
@@ -1025,34 +1025,64 @@ Random Number Generation for Probabilistic Public Key Algorithms
...
+
+   **Output:**
+
+   -  :math:`r`: :math:`min \leq r < max`
+
+   **Steps:**
+
+   1. Preliminary parameter requirement checks are conducted. :math:`min` must be
+      smaller than :math:`max`, and neither can be negative.
+   2. Calculate :math:`diff` as the difference between :math:`max` and :math:`min`
```
Comment (FAlbertDev) [inline docs/cryptodoc/src/07_rng.rst:1083]: _diff_ in math font looks weird, but I don't know why. It's okay, I guess...

image

**Review by FAlbertDev — APPROVED** [review]
LGTM

File: docs/cryptodoc/src/07_rng.rst (hunk 6 above)
Comment (lieser) [reply docs/cryptodoc/src/07_rng.rst:1083]: It is more by chance I think that the rest looks ok. I think in math mode it is more assumed that each letter is it's own variable. All multi letter variable names need probably to be in a `\mathit` or `\mathrm`.

See also https://tex.stackexchange.com/questions/482743/typesetting-multi-letter-variable-names-in-math-mode

Comment (reneme) [reply docs/cryptodoc/src/07_rng.rst:1083]: Yikes! That makes sense, I guess. I'll see what I can do.

Comment (reneme) [reply docs/cryptodoc/src/07_rng.rst:1083]: Mhh, I didn't find a convenient way to express this in Sphinx. I guess, for now, we have to live with this crappy kerning. 😞
One day we should donate a crate of beer for the one willing to go through the whole document and rectify all the formatting quirks.

**Pull request merged**